File: backend/internal/handler/collection_handler.go
package handler

import (
	"net/http"
	"strconv"

	"github.com/gin-gonic/gin"
	"github.com/kenyamamoto/pokelingual/backend/internal/domain"
	"github.com/kenyamamoto/pokelingual/backend/internal/service"
)

// CollectionHandler handles HTTP requests for the Pokemon collection.
type CollectionHandler struct {
	collectionService *service.CollectionService
	settingsRepo      domain.UserSettingsRepository
}

// NewCollectionHandler creates a new CollectionHandler.
func NewCollectionHandler(collectionService *service.CollectionService, settingsRepo domain.UserSettingsRepository) *CollectionHandler {
	return &CollectionHandler{
		collectionService: collectionService,
		settingsRepo:      settingsRepo,
	}
}

// GetCollection handles GET /collection to list the user's discovered Pokemon.
func (h *CollectionHandler) GetCollection(c *gin.Context) {
	uid := c.GetString("uid")

	entries, err := h.collectionService.GetCollection(c.Request.Context(), uid)
	if err != nil {
		handleError(c, err)
		return
	}

	totalAvailable := service.MaxPokemonID

	capturedCount := 0
	for _, e := range entries {
		if e.Status == "captured" {
			capturedCount++
		}
	}

	c.JSON(http.StatusOK, gin.H{
		"pokemon":         entries,
		"total_available": totalAvailable,
		"captured_count":  capturedCount,
	})
}

// GetPokemonDetail handles GET /collection/:id to get details of a captured Pokemon.
func (h *CollectionHandler) GetPokemonDetail(c *gin.Context) {
	uid := c.GetString("uid")

	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid pokemon id"})
		return
	}

	detail, err := h.collectionService.GetPokemonDetail(c.Request.Context(), uid, id)
	if err != nil {
		handleError(c, err)
		return
	}

	c.JSON(http.StatusOK, detail)
}

File: backend/src/handler/collection-handler.ts
import type { Request, Response } from "express";
import type { PokemonConfig, UserSettingsRepository } from "../domain/ports.js";
import type { CollectionService } from "../service/collection-service.js";
import type { CollectionResponse } from "../../../shared/api-types/collection.js";
import { handleError } from "./error.js";

/** ポケモン図鑑 (コレクション) 取得用エンドポイントを束ねるハンドラ。 */
export class CollectionHandler {
  /**
   * @param collectionService 図鑑取得のドメインサービス。
   * @param settingsRepo ユーザ設定リポジトリ。
   * @param pokemonConfig ポケモン関連のアプリ設定。
   */
  constructor(
    private collectionService: CollectionService,
    private settingsRepo: UserSettingsRepository,
    private pokemonConfig: PokemonConfig,
  ) {}

  /**
   * GET /collection — ユーザの図鑑一覧を返す。
   * @param req Express リクエスト。
   * @param res Express レスポンス。
   */
  getCollection = async (req: Request, res: Response) => {
    const uid = res.locals.uid as string;
    try {
      const { entries, unavailable_count } = await this.collectionService.getCollection(uid);
      const capturedCount = entries.filter((e) => e.status === "captured").length;
      const body: CollectionResponse = {
        pokemon: entries,
        total_available: this.pokemonConfig.maxPokemonID,
        captured_count: capturedCount,
        unavailable_count,
      };
      res.json(body);
    } catch (err) {
      handleError(res, err, req.path);
    }
  };

  /**
   * GET /collection/:id — 特定ポケモンの詳細を返す。
   * @param req Express リクエスト。
   * @param res Express レスポンス。
   */
  getPokemonDetail = async (req: Request, res: Response) => {
    const uid = res.locals.uid as string;
    const id = parseInt(String(req.params.id), 10);
    if (isNaN(id)) {
      res.status(400).json({ error: "invalid pokemon id" });
      return;
    }
    try {
      const detail = await this.collectionService.getPokemonDetail(uid, id);
      res.json(detail);
    } catch (err) {
      handleError(res, err, req.path);
    }
  };
}
